src/methods/abstract.js

```diff
@@ -74,6 +75,26 @@ export function RequireObjectCoercible(
   argLoc?: ?BabelNodeSourceLocation
 ): AbstractValue | ObjectValue | BooleanValue | StringValue | SymbolValue | NumberValue {
   if (arg instanceof AbstractValue && (arg.mightBeNull() || arg.mightBeUndefined())) {
+    if (realm.isInPureScope()) {
+      if (realm.isInPureTryStatement) {
+        // TODO(1264): We should be able to preserve error handling on abstract throw
+        // but currently we just issue a recoverable error instead.
+        let diag = new CompilerDiagnostic(
+          "Possibly throwing object coercion inside try/catch",
+          argLoc || realm.currentLocation,
+          "PP0021",
+          "RecoverableError"
+        );
+        if (realm.handleError(diag) !== "Recover") throw new FatalError();
+      }
+      // In a pure function it is ok to throw if this happens to be null or undefined.
+      // So we can safely coerce this to an abstract object value.
+      // We don't actually have to do anything since any op that forced this to be an
+      // object will do that for us.
+      let builder = preludeGenerator => ({ A }: any) => t.sequenceExpression([A]);
```

Comment:
I am befuddled by this. It seems very different from other places where createFromTemplate is used.

Reply:
Usually it is used with some kind of operation described in source.

https://github.com/facebook/prepack/blob/f7b0cb8ecd099e5fc5ef8a8b6aff8bf791664782/src/intrinsics/ecma262/BooleanPrototype.js#L21-L22

However, this doesn't have an equivalent operation since it'll be implicitly called. I just need this to reference the original value.

I could just return the original expression but the problem is this invariant:

https://github.com/facebook/prepack/blob/master/src/serializer/ResidualHeapSerializer.js#L1483-L1486

It makes too strong assumption about what an identifier as the top node means. It is still useful to have though and I don't really want to refactor that whole thing.

So I wrapped it in a sequence expression to workaround that issue. Creating a sequence expression from source seemed very fragile though since the parser can easily remove it when there is only one expression.

Reply:
Hmm. AbstractValue.createFromBuildFunction (which does not currently exist) would be less surprising. Just returning the object, as is, seems like the obviously correct thing to do here, however.

What seems to be happening here is that arg is further abstracted by setting its ValuesDomain to top. The rest is just smoke and mirrors. It is probably interesting to find out why this is necessary. At the very least it should be much more obvious that is happening (preferably with a source code comment explaining why it is needed).

Reply:
The primary goal of this is to set the TypesDomain to one compatible with ObjectValue. That in turn causes this to be an AbstractObjectValue which makes the rest work.

As a consequence the ValuesDomain becomes top but it probably doesn't have to be. It could be something more precise.

Reply:
Ah yes, you are quite right, this is all about setting the TypeDomain to ObjectValue. Well then, what is need here is an obvious way to do just that.

Reply:
Ok I made a new `createFromCoercionOp`.